services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java

```diff
@@ -103,17 +115,68 @@ public String encode(Token token) {
     }
 
     @Override
-    public <T> T decodeClientJWT(String token, ClientModel client, Class<T> clazz) {
-        if (token == null) {
+    public <T> T decodeClientJWT(String jwt, ClientModel client, BiConsumer<JOSE, ClientModel> jwtValidator, Class<T> clazz) {
+        if (jwt == null) {
             return null;
         }
-        try {
-            JWSInput jws = new JWSInput(token);
 
-            String signatureAlgorithm = jws.getHeader().getAlgorithm().name();
+        JOSE joseToken = JOSEParser.parse(jwt);
+
+        jwtValidator.accept(joseToken, client);
+
+        if (joseToken instanceof JWE) {
+            try {
+                Optional<KeyWrapper> activeKey;
+                String kid = joseToken.getHeader().getKeyId();
+                Stream<KeyWrapper> keys = session.keys().getKeysStream(session.getContext().getRealm());
+
+                if (kid == null) {
+                    activeKey = keys.filter(k -> KeyUse.ENC.equals(k.getUse()) && k.getPublicKey() != null)
+                            .sorted(Comparator.comparingLong(KeyWrapper::getProviderPriority).reversed())
+                            .findFirst();
+                } else {
+                    activeKey = keys.filter(k -> k.getKid().equals(kid)).findAny();
```

Comment:
Even if "kid" is not null, will it be safer to still check if the key usage is ENC? To make sure that the "kid" used doesn't come from the key, which is supposed to be used for SIG?